File: backend/app/schemas/menu_filters.py
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

MenuSortOption = Literal["default", "price_asc", "price_desc", "name"]


class MenuItemFilterParams(BaseModel):
    category_id: int | None = None
    available_only: bool = False
    featured_only: bool = False
    dietary_tags: list[str] = Field(default_factory=list)
    max_price: Decimal | None = None
    min_price: Decimal | None = None
    search: str | None = None
    sort: MenuSortOption = "default"


class DietaryTagFacet(BaseModel):
    tag: str
    label: str
    count: int


class MenuPriceRange(BaseModel):
    min: Decimal
    max: Decimal


class MenuFilterFacetsRead(BaseModel):
    dietary_tags: list[DietaryTagFacet]
    price_range: MenuPriceRange
    featured_count: int
    available_count: int
    total_count: int

File: src/types/menuFilters.ts
export type DietaryTag = 'VEGETARIAN' | 'VEGAN' | 'GLUTEN_FREE' | 'SPICY';

export type MenuSortOption = 'default' | 'price_asc' | 'price_desc' | 'name';

export interface MenuItemFilters {
  availableOnly: boolean;
  featuredOnly: boolean;
  dietaryTags: DietaryTag[];
  maxPrice: number | null;
  sort: MenuSortOption;
}

export interface DietaryTagFacet {
  tag: DietaryTag;
  label: string;
  count: number;
}

export interface MenuFilterFacets {
  dietary_tags: DietaryTagFacet[];
  price_range: { min: number; max: number };
  featured_count: number;
  available_count: number;
  total_count: number;
}

export const DEFAULT_MENU_FILTERS: MenuItemFilters = {
  availableOnly: false,
  featuredOnly: false,
  dietaryTags: [],
  maxPrice: null,
  sort: 'default',
};

export const DIETARY_TAG_LABELS: Record<DietaryTag, string> = {
  VEGETARIAN: 'Vegetarian',
  VEGAN: 'Vegan',
  GLUTEN_FREE: 'Gluten-free',
  SPICY: 'Spicy',
};
